core/src/main/java/hudson/model/AbstractBuild.java

```diff
@@ -518,6 +559,20 @@ public Result run(@Nonnull BuildListener listener) throws Exception {
             return result;
         }
 
+        /**
+         * An {@link Environment} which does nothing, but change state when it gets torn down. Used in
+         * {@link AbstractBuildExecution#run(BuildListener)} to detect whether environments have yet to be torn down,
+         * or if it has been done already (in the {@link AbstractBuildExecution#doRun(BuildListener)} implementation).
+         */
+        private class TearDownCheckEnvironment extends Environment {
```

Comment:
```suggestion
        private static class TearDownCheckEnvironment extends Environment {
```

Reply:
Actually, now I remember... It's not static because it is nested in a non-static nested class (`AbstractBuild.AbstractBuildExecution`). Do you think I should move it?

Reply:
Would be better to avoid non-static nested classes when possible. Better safe than sorry

Reply:
Okay,that's 462104a